csrc/multidevice/allocator.cpp

```diff
@@ -0,0 +1,112 @@
+copyFusionAndChangeOutputs(Fusion* fusion, std::unordered_set<Val*> outputs) {
+  std::unique_ptr<Fusion> fusion_copy = std::make_unique<Fusion>();
+  std::unordered_map<Val*, Val*> copy_to_original_map;
+  auto original_to_copy_cloner = Fusion::copy(fusion, fusion_copy.get());
+
+  auto original_inputs = fusion_copy->inputs();
+  auto original_outputs = fusion_copy->outputs();
+
+  // Remove original outputs
+  std::for_each(
+      original_outputs.begin(), original_outputs.end(), [&](auto& output) {
+        fusion_copy->removeOutput(output);
+      });
+
+  // Add new outputs
+  std::for_each(outputs.begin(), outputs.end(), [&](Val* const& output) {
+    fusion_copy->addOutput(original_to_copy_cloner.clone(output));
+    copy_to_original_map[original_to_copy_cloner.clone(output)] = output;
+  });
+
+  for (auto tv : ir_utils::filterByType<TensorView>(fusion_copy->vals())) {
+    tv->setMemoryType(MemoryType::Global);
+    for (auto i : c10::irange(tv->domain()->nDims())) {
+      tv->axis(static_cast<int>(i))->parallelize(ParallelType::Serial);
+    }
+  }
+
+  return std::
+      make_pair<std::unique_ptr<Fusion>, std::unordered_map<Val*, Val*>>(
+          std::move(fusion_copy), std::move(copy_to_original_map));
+}
+
+} // namespace
+
+std::unordered_map<Val*, c10::IValue> allocatePipelineIntermediateBuffers(
+    Pipeline* pipeline,
+    DeviceIdxType my_device_index,
+    std::vector<c10::IValue> global_inputs_IValues) {
+  // Stores the Vals that needs to be allocated
+  std::unordered_set<Val*> vals_to_allocate;
+  // Add all the input of stages run by the current device
+  const auto& exprs = pipeline->exprs();
+  for (auto stage : ir_utils::filterByType<PipelineStage>(exprs)) {
+    if (stage->descriptor()->mesh.has(my_device_index)) {
+      for (auto input : stage->inputs()) {
+        auto input_val = input->as<PipelineVal>()->getOriginalVal();
+        vals_to_allocate.insert(input_val);
+      }
+    }
+  }
+  // Remove any global inputs that have been added
+  for (auto global_input : pipeline->originalFusion()->inputs()) {
+    vals_to_allocate.erase(global_input);
+  }
+  // Remove any intermediate stage's output already produced
+  for (auto stage : ir_utils::filterByType<PipelineStage>(exprs)) {
+    if (stage->descriptor()->mesh.has(my_device_index)) {
+      for (auto output : stage->outputs()) {
+        auto output_val = output->as<PipelineVal>()->getOriginalVal();
+        vals_to_allocate.erase(output_val);
+      }
+    }
+  }
+
+  // We copy the original fusion and set the outputs to be the tensors to be
+  // allocated. This way we can directly use FusionExecutor::allocOutputSpace
+  auto [fusion_copy, copy_to_original_map] =
+      copyFusionAndChangeOutputs(pipeline->originalFusion(), vals_to_allocate);
+  if (fusion_copy->outputs().empty()) {
+    return {};
+  }
+  FusionExecutor fe;
+  fe.compileFusion(fusion_copy.get(), global_inputs_IValues);
```

Comment:
This means the copied fusion is lowered and compiled by nvrtc. And the compiled kernel is just discarded. Why do we need to do that just to allocate interemediates?

Reply:
It is just the easiest way I found to allocate a given set of tensors. Do you have another suggestion?

Reply:
I suppose there's a way to allocate outputs without compiling the kernel and it should be just a matter of factoring out the allocation interface to work with a `Fusion` rather than a `Kernel`.

Reply:
I'm not sure how cumbersome the refactoring would be, but compiling a fusion just for allocating outputs seems to be quite wasteful.

Reply:
I think the refactoring will be substantial and belongs to another PR. There is also a need to change the `FusionExecutor` interface since the `compile` method is the only way I see (except `runFusion`) to set the Fusion.
In the meantime I really see no other option than doing like that. Let me know what you suggest